pyomo/contrib/appsi/cmodel/src/expression.cpp

```diff
@@ -1,478 +1,328 @@
-  else if (other.is_constant_type() && other.evaluate() == 0.0)
+  else if (domain == "binary")
     {
-      return other.shared_from_this();
+      lb2 = 0;
     }
-  else if (is_constant_type() && evaluate() == 1.0)
+  else if (domain == "percent_fraction")
     {
-      return other.shared_from_this();
+      lb2 = 0;
     }
-  else if (is_constant_type() && evaluate() == 0.0)
+  else if (domain == "unit_interval")
     {
-      return shared_from_this();
+      lb2 = 0;
     }
   else
     {
-      std::shared_ptr<MultiplyOperator> oper = std::make_shared<MultiplyOperator>();
-      return binary_helper(shared_from_this(), other.shared_from_this(), oper);
+      throw py::value_error("Unrecognized domain: " + domain);
     }
+  return std::max(lb1, lb2);
 }
 
 
-std::shared_ptr<ExpressionBase> ExpressionBase::operator/(ExpressionBase& other)
+double Var::get_ub()
 {
-  if (other.is_constant_type() && other.evaluate() == 1.0)
+  double ub1 = ub->evaluate();
+  double ub2;
+  if (domain == "reals")
     {
-      return shared_from_this();
+      ub2 = inf;
     }
-  else if (other.is_constant_type() && other.evaluate() == 0.0)
+  else if (domain == "nonnegative_reals")
     {
-      assert (false);
+      ub2 = inf;
     }
-  else if (is_constant_type() && evaluate() == 0.0)
+  else if (domain == "nonpositive_reals")
     {
-      return shared_from_this();
+      ub2 = 0;
     }
-  std::shared_ptr<DivideOperator> oper = std::make_shared<DivideOperator>();
-  return binary_helper(shared_from_this(), other.shared_from_this(), oper);
-}
-
-
-std::shared_ptr<ExpressionBase> ExpressionBase::__pow__(ExpressionBase& other)
-{
-  if (other.is_constant_type() && other.evaluate() == 1.0)
+  else if (domain == "integers")
     {
-      return shared_from_this();
+      ub2 = inf;
     }
-  else if (other.is_constant_type() && other.evaluate() == 0.0)
+  else if (domain == "nonnegative_integers")
     {
-      return std::make_shared<Constant>(1.0);
+      ub2 = inf;
     }
-  else if (is_constant_type() && evaluate() == 1.0)
+  else if (domain == "nonpositive_integers")
     {
-      return shared_from_this();
+      ub2 = 0;
     }
-  else if (is_constant_type() && evaluate() == 0.0)
+  else if (domain == "binary")
     {
-      return shared_from_this();
+      ub2 = 1;
     }
-  else
+  else if (domain == "percent_fraction")
     {
-      std::shared_ptr<PowerOperator> oper = std::make_shared<PowerOperator>();
-      return binary_helper(shared_from_this(), other.shared_from_this(), oper);
+      ub2 = 1;
     }
-}
-
-
-std::shared_ptr<ExpressionBase> ExpressionBase::operator-()
-{
-  if (is_constant_type())
+  else if (domain == "unit_interval")
     {
-      return std::make_shared<Constant>(-evaluate());
+      ub2 = 1;
     }
   else
     {
-      std::shared_ptr<NegationOperator> oper = std::make_shared<NegationOperator>();
-      return unary_helper(shared_from_this(), oper);
+      throw py::value_error("Unrecognized domain: " + domain);

```

Comment:
The problem is that I don't have a C++ representation of a Set, and I don't have access to the Pyomo Set at this point.

Reply:
@jsiirola I think I have done something reasonable now. In order for the automatic updates to work correctly, I have to store the bounds and the domain completely separate. However, I was able to make use of `get_interval`. I just store the results from that instead of the `domain`. I also added an enum to get rid of the strings.